sdk/identity/Azure.Identity/tests/AuthenticationRecordTests.cs

```diff
@@ -0,0 +1,104 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.IO;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using NUnit.Framework;
+
+namespace Azure.Identity.Tests
+{
+    public class AuthenticationRecordTests
+    {
+        [Test]
+        public void SerializeDeserializeInputChecks()
+        {
+            var record = new AuthenticationRecord(Guid.NewGuid().ToString(), Guid.NewGuid().ToString(), Guid.NewGuid().ToString(), Guid.NewGuid().ToString());
+
+            Assert.Throws<ArgumentNullException>(() => record.Serialize(null));
+            Assert.ThrowsAsync<ArgumentNullException>(async () => await record.SerializeAsync(null));
+            Assert.Throws<ArgumentNullException>(() => AuthenticationRecord.Deserialize(null));
+            Assert.ThrowsAsync<ArgumentNullException>(async () => await AuthenticationRecord.DeserializeAsync(null));
+        }
+
+        [Test]
+        public async Task SerializeDeserializeAsync()
+        {
+            var expRecord = new AuthenticationRecord(Guid.NewGuid().ToString(), Guid.NewGuid().ToString(), Guid.NewGuid().ToString(), Guid.NewGuid().ToString());
+
+            byte[] buff = new byte[256];
+
+            var stream = new MemoryStream(buff);
+
+            await expRecord.SerializeAsync(stream);
+
+            // decode to string and re-encode to strip extra bytes
+            var buffStr = Encoding.UTF8.GetString(buff).Trim('\0');
+
+            buff = Encoding.UTF8.GetBytes(buffStr);
+
+            stream = new MemoryStream(buff);
+
+            var actRecord = await AuthenticationRecord.DeserializeAsync(stream);
+
+            Assert.AreEqual(expRecord.Username, actRecord.Username);
+            Assert.AreEqual(expRecord.Authority, actRecord.Authority);
+            Assert.AreEqual(expRecord.HomeAccountId, actRecord.HomeAccountId);
+            Assert.AreEqual(expRecord.TenantId, actRecord.TenantId);
+        }
+
+        [Test]
+        public void SerializeDeserialize()
+        {
+            var expRecord = new AuthenticationRecord(Guid.NewGuid().ToString(), Guid.NewGuid().ToString(), Guid.NewGuid().ToString(), Guid.NewGuid().ToString());
+
+            byte[] buff = new byte[256];
+
+            var stream = new MemoryStream(buff);
+
+            expRecord.Serialize(stream);
+
+            // decode to string and re-encode to strip extra bytes
+            var buffStr = Encoding.UTF8.GetString(buff).Trim('\0');
+
+            buff = Encoding.UTF8.GetBytes(buffStr);
+
+            stream = new MemoryStream(buff);
```

Comment:
nit: you can seek memory stream.

Reply:
I needed to trim down to just the written bytes, but I did this with your suggestion of using the current position of the stream after being read from to create a new stream of the exact size.